### ISSUE TITLE
trkid= not cleaned from Netflix links when codex.nflxext.com is blocked

### ISSUE THREAD

**Issue by vertigo220** [KevinRoebert/ClearUrls#40]
### Prerequisites
* [Yes, with the exception of uBo and uMatrix] Can you reproduce the problem in a **fresh** firefox profile with **only** ClearURLs?
* [Yes] Are you running the latest version of ClearURLs?
* [Yes & No] Are you have changed settings under `about:config`?

### Description

Not sure if this is a bug or not, but codex.nflxext.com script has to be allowed in order for a trkid Netflix link to be cleaned. For example, with uBo and uMatrix installed, going to https://www.netflix.com/title/80166314?trkid=1234567&MSG_TITLE=80166314&lnktrk=EMP&g=6E5416B8D93DD75DB84131269FB4725A6455A8C2&lkid=W2W_ROW_1_MDP_ will only convert the link to https://www.netflix.com/title/80166314 once it has been allowed. It seems the conversion shouldn't be reliant on requiring script to run on the destination page.

I was going to make a separate issue for the fact that google search results aren't cleaned by ClearURLs, but I now suspect it's likely due to the same cause. It seems very counter-productive to globally require 3rd-party scripts just to clean links, so I'm hoping this is just a bug and not a requirement for it to function.

### Steps to Reproduce

1. Install uBo and uMatrix and set to block 3rd-party (alternately, you could just use uBo and forego uMatrix, as the result would be the same, since uMatrix basically just allows more fine-grained control)
2. Go to above link and notice it's not cleaned
3. Allow the above in uBo and, if applicable, allow scripts for it in uMatrix, then go to the link again and notice that it's cleaned

**Expected behavior:** [What you expected to happen]

The link should be cleaned regardless of what scripts are allowed to run on the site

**Actual behavior:** [What actually happened]

Netflix links (and others?) aren't cleaned until certain 3rd-party scripts are allowed/unblocked.

### System
* **Version**: 1.2.1.4
* **Platform**: Win10x64

### Log

There were no log entries created with the script was blocked. Both entries are from loading the page a single time after unblocking it. Also, I've noticed many entries have the "This domain is blocked" statement in the rule column.

| Before processing | After processing | Rule | Time |
|-------------------|------------------|------|------|
|                   |                  |      |      |

https://adtech.nflximg.net/adtech_iframe_target_03.html?data=%7B%22membership_status%22%3A%22NON_REGISTERED_MEMBER%22%2C%22country%22%3A%22US%22%2C%22source%22%3A%22https%3A%2F%2Fwww.netflix.com%2Ftitle%2F80166314%3Ftrkid%3D13710079%26amp%3BMSG_TITLE%3D80166314%26amp%3Blnktrk%3DEMP%26amp%3Bg%3D3E5616B8D73DD75DB84131269FB4725A6455A8C2%26amp%3Blkid%3DW2W_ROW_1_MDP_1%22%2C%22fbaId%22%3A%226ef0941b-6a2f-47a9-89db-e12565dc2d3e%22%2C%22is_member%22%3A%22anonymous%22%2C%22wasFormerMember%22%3Afalse%2C%22referrer%22%3A%22title%22%7D |   | .* | 5/8/2018, 6:31:24 AM
-- | -- | -- | --
https://adtech.nflximg.net/adtech_iframe_target_03.html?data=%7B%22membership_status%22%3A%22NON_REGISTERED_MEMBER%22%2C%22country%22%3A%22US%22%2C%22source%22%3A%22https%3A%2F%2Fwww.netflix.com%2Ftitle%2F80166314%3Ftrkid%3D13710079%26amp%3BMSG_TITLE%3D80166314%26amp%3Blnktrk%3DEMP%26amp%3Bg%3D3E5616B8D73DD75DB84131269FB4725A6455A8C2%26amp%3Blkid%3DW2W_ROW_1_MDP_1%22%2C%22fbaId%22%3A%226ef0941b-6a2f-47a9-89db-e12565dc2d3e%22%2C%22is_member%22%3A%22anonymous%22%2C%22wasFormerMember%22%3Afalse%2C%22referrer%22%3A%22title%22%7D | https://adtech.nflximg.net/adtech_iframe_target_03.html?data=%7B%22membership_status%22%3A%22NON_REGISTERED_MEMBER%22%2C%22country%22%3A%22US%22%2C%22source%22%3A%22https%3A%2F%2Fwww.netflix.com%2Ftitle%2F80166314%3Ftrkid%3D13710079%26amp%3BMSG_TITLE%3D80166314%26amp%3Blnktrk%3DEMP%26amp%3Bg%3D3E5616B8D73DD75DB84131269FB4725A6455A8C2%26amp%3Blkid%3DW2W_ROW_1_MDP_1%22%2C%22fbaId%22%3A%226ef0941b-6a2f-47a9-89db-e12565dc2d3e%22%2C%22is_member%22%3A%22anonymous%22%2C%22wasFormerMember%22%3Afalse%2C%22referrer%22%3A%22title%22%7D | This domain is blocked. | 5/8/2018, 6:31:24 AM

**Comment by KevinRoebert** [comment]
With call I mean a request to the server. ClearURLs cleans only those requests that are actually sent to a server.

There are also links on pages which are never called, e.g. in the Google search, you only call the links you are clicked on.

Something similar you have to imagine with the Netflix link. It does contain tracking fields, but it's never called by Netflix, if you're blocking the 3rd-party scripts.

Netflix and other vendors sometimes add these tracking fields after the request, so as to have e.g. bypassing addons like ClearURLs. You will only be able to track people who have not installed a privacy addon such as ClearURLs. But you will be protected from this trick (I've tried to show you the trick with my example page on curl.kevinroebert.de), if you have installed ClearURLs or something similar.

As a small proof that no request, even when clicking on the Netflix video is made, I have attached a screenshot. There I explicitly searched for `netflix.com/watch`:

![prof_netflix_no_request](https://user-images.githubusercontent.com/28713602/39776690-da560036-5301-11e8-8888-81b548b22132.PNG)

The situation is different, however, if you reload the video with F5, then your browser will make a request to Netflix with the tracking fields. Then ClearURLs detects this request and deletes the tracking fields before submitting.

We purposely designed ClearURLs this way to purify the URLs when they actually call. Thus, we save a lot of computing power, because otherwise many unnecessary links would be cleaned up, but never visited or loaded by you.

To summarize in a nutshell: ClearURLs cleans up all the urls that are called by your browser, so no tracking field that could detect ClearURLs ever arrives at the provider of the page.

**Comment by vertigo220** [comment]
Ok, I'm pretty sure I'm more or less understanding this, but the concern I have is that regardless of whether this addon is enabled or not, everything acts exactly the same. Hovering over links in a google search show the clean link, then clicking on them causes them to change to the tracking link, which is what you get if you copy it, but then when it loads the page the URL in the location bar is clean. Since it's like this with or without the addon, it doesn't seem like it's doing anything, even if it is in the background, and so I never know if it's actually working or not aside from looking at the logs, and I hate to just assume something is working. I take it though that's just how it is with this and I just have to have faith that it's doing its job?

**Comment by KevinRoebert** [comment]
If you hovering over a url, does not mean you see the real link. It can be changed by a script. The real url is only final after a press on it.

Only at this point (the request) ClearURLs and I think NeatURLs too, clean up the request and not the shown url, from tracking fields. You can see all actions from ClearURLs in the log file and you can see the "blocked" counter growing. So you do not have to trust me, you can see a proof of work in the log and at the "blocked" counter.

I believe the real problem is an understanding problem. You must not assume that urls that you see when you click are also called. Urls can be manipulated in the background of scripts and this also happens with Netflix and Google, for example through the `history.pushState` function.

A url is always final and invariable only at the request to the server. This is where ClearURLs comes in and cleans up the request for tracking fields. You may not be able to see the cleaned url on your monitor, but you can see in the log files or at the counter that the request has been cleaned up.

**Comment by vertigo220** [comment]
> If you hovering over a url, does not mean you see the real link. It can be changed by a script. The real url is only final after a press on it.

I understand this. I think I've found the reason for much of my confusion, though. I was wondering why, when clicking the links, I was seeing all the tracking info, but there was no indication it was being cleaned. Even using the badge and log, I wasn't seeing it. Turns out, it was because I have the Do Not Track setting enabled and, somewhat surprisingly, that's apparently honored by Google. I disabled that and retested a link and, sure enough, the ClearURLs log showed the link being cleaned as expected. What got me to think to try that was your statement "You must not assume that urls that you see when you click are also called," which got me wondering why bother embed the tracking info if they're not going to use it. Now things make a lot more sense. Thanks for the help :)
